File: Comportamentales Java/ChainOfResponsibility.java
// Chain of Responsibility Pattern
abstract class Handler {
    private Handler next;
    
    public Handler setNext(Handler handler) {
        this.next = handler;
        return handler;
    }
    
    public abstract String handle(String request);
    
    protected String passToNext(String request) {
        if (next != null) {
            return next.handle(request);
        }
        return null;
    }
}

class MonkeyHandler extends Handler {
    @Override
    public String handle(String request) {
        if ("Banana".equals(request)) {
            return "Monkey: I'll eat the " + request;
        }
        return passToNext(request);
    }
}

class SquirrelHandler extends Handler {
    @Override
    public String handle(String request) {
        if ("Nut".equals(request)) {
            return "Squirrel: I'll eat the " + request;
        }
        return passToNext(request);
    }
}

class DogHandler extends Handler {
    @Override
    public String handle(String request) {
        if ("MeatBall".equals(request)) {
            return "Dog: I'll eat the " + request;
        }
        return passToNext(request);
    }
}

// Usage
class ChainOfResponsibilityDemo {
    public static void main(String[] args) {
        Handler monkey = new MonkeyHandler();
        Handler squirrel = new SquirrelHandler();
        Handler dog = new DogHandler();
        
        monkey.setNext(squirrel).setNext(dog);
        
        String[] foods = {"Nut", "Banana", "Cup of coffee"};
        for (String food : foods) {
            String result = monkey.handle(food);
            if (result != null) {
                System.out.println(result);
            } else {
                System.out.println("  " + food + " was left untouched.");
            }
        }
    }
}

File: Comportamentales Python/chain_of_responsibility.py
# Chain of Responsibility Pattern
from abc import ABC, abstractmethod

class Handler(ABC):
    @abstractmethod
    def set_next(self, handler):
        pass
    
    @abstractmethod
    def handle(self, request):
        pass

class AbstractHandler(Handler):
    _next_handler = None
    
    def set_next(self, handler: Handler):
        self._next_handler = handler
        return handler
    
    def handle(self, request):
        if self._next_handler:
            return self._next_handler.handle(request)
        return None

class MonkeyHandler(AbstractHandler):
    def handle(self, request):
        if request == "Banana":
            return f"Monkey: I'll eat the {request}"
        else:
            return super().handle(request)

class SquirrelHandler(AbstractHandler):
    def handle(self, request):
        if request == "Nut":
            return f"Squirrel: I'll eat the {request}"
        else:
            return super().handle(request)

class DogHandler(AbstractHandler):
    def handle(self, request):
        if request == "MeatBall":
            return f"Dog: I'll eat the {request}"
        else:
            return super().handle(request)

# Usage
if __name__ == "__main__":
    monkey = MonkeyHandler()
    squirrel = SquirrelHandler()
    dog = DogHandler()
    
    monkey.set_next(squirrel).set_next(dog)
    
    for food in ["Nut", "Banana", "Cup of coffee"]:
        result = monkey.handle(food)
        if result:
            print(result)
        else:
            print(f"  {food} was left untouched.")
